waveform_editor/derived_waveform.py

```diff
@@ -0,0 +1,170 @@
+
+from waveform_editor.base_waveform import BaseWaveform
+
+
+class DependencyRenamer(ast.NodeTransformer):
+    """AST transformer to rename string constants."""
+
+    def __init__(self, rename_from, rename_to, yaml):
+        self.rename_from = rename_from
+        self.rename_to = rename_to
+        self.yaml = yaml
+
+    def visit_Constant(self, node):
+        """
+        Replace string constants equal to `rename_from` with `rename_to`
+        and update the YAML source lines accordingly.
+        """
+        if isinstance(node.value, str) and node.value == self.rename_from:
+            split_yaml = self.yaml.splitlines()
+            line_number = node.lineno - 1
+            line = split_yaml[line_number]
+            split_yaml[line_number] = (
+                line[: node.col_offset]
+                + line[node.col_offset : node.end_col_offset].replace(
+                    self.rename_from, self.rename_to
+                )
+                + line[node.end_col_offset :]
+            )
+            self.yaml = "\n".join(split_yaml)
+            return ast.copy_location(ast.Constant(value=self.rename_to), node)
+        return node
+
+
+class ExpressionExtractor(ast.NodeTransformer):
+    """
+    AST transformer extracting all string constants from expressions
+    and replacing them with Name nodes for later evaluation.
+    """
+
+    def __init__(self):
+        self.string_nodes = []
+        self.is_constant = True
+
+    def visit_Constant(self, node):
+        if isinstance(node.value, str):
+            self.string_nodes.append(node.value)
+            self.is_constant = False
+            return ast.copy_location(
+                ast.Subscript(
+                    value=ast.Name(id="__w", ctx=ast.Load()),
+                    slice=ast.Constant(value=node.value),
+                    ctx=ast.Load(),
+                ),
+                node,
+            )
+        elif isinstance(node.value, (int, float)):
+            return node
+        else:
+            self.is_constant = False
+            return node
+
+
+class DerivedWaveform(BaseWaveform):
+    def __init__(self, yaml_str, name, config, dd_version=None):
+        super().__init__(yaml_str, name, dd_version)
+        self.config = config
+        self.dependent_waveforms = set()
+        self.is_constant = False
+        self.expression = None
+        self.prepare_expression()
+
+    def prepare_expression(self):
+        """Parse the YAML expression, extract dependencies, transform it for
+        evaluation, and compile it.
+        """
+        if self.yaml is None:
+            return
+
+        try:
+            tree = ast.parse(str(self.yaml), mode="eval")
+        except Exception as e:
+            self.annotations.add(0, f"Could not parse or evaluate the waveform: {e}")
+            self.expression = None
+            return
+
+        extractor = ExpressionExtractor()
+        modified_tree = extractor.visit(ast.fix_missing_locations(tree))
```

Comment:
This should be executed the other way around, right? https://docs.python.org/3/library/ast.html#ast.NodeTransformer

```suggestion
        modified_tree = ast.fix_missing_locations(extractor.visit(tree))
```